src/components/node-list/node-list-items.js

```diff
@@ -330,3 +349,314 @@ export const getFilteredItems = createSelector(
+          filteredMoudularPipelineParents.push(
+            // construct the item needed and then add it to the list
+            constructModularPipelineItem(
+              modularPipelines.filter(
+                (rawModularPipeline) => rawModularPipeline.id === parent
+              )[0]
+            )
+          );
+        }
+        lastIndex = childrenModularPipeline.id.indexOf('.', lastIndex) + 1;
+      }
+    });
+
+    return filteredMoudularPipelineParents;
+  }
+);
+
+/**
+ * returns the corresponding final set of modular pipelines
+ * for constructing the final nested tree list
+ */
+export const getFilteredTreeItems = createSelector(
+  [
+    getFilteredModularPipelineItems,
+    getFilteredNodeModularPipelines,
+    getFilteredModularPipelineParent,
+    (state) => state.modularPipelines,
+  ],
+  (
+    modularPipelineItems,
+    nodeModularPipelines,
+    modularPipelineParent,
+    modularPipelines
+  ) => {
+    modularPipelineItems = modularPipelineItems.modularPipeline;
+
+    let finalModularPipelines = [];
+
+    // sort all 3 sets of modular pipelines according to the original order
+    modularPipelines?.forEach((modularPipeline) => {
+      // check filtered modular pipelines
+      if (
+        modularPipelineItems.some(
+          (modularPipelineItem) => modularPipelineItem.id === modularPipeline.id
+        )
+      ) {
+        finalModularPipelines.push(
+          modularPipelineItems.filter(
+            (modularPipelineItem) =>
+              modularPipelineItem.id === modularPipeline.id
+          )[0]
+        );
+      } else if (
+        nodeModularPipelines.some(
+          (nodeModularPipeline) => nodeModularPipeline.id === modularPipeline.id
+        )
+      ) {
+        finalModularPipelines.push(
+          nodeModularPipelines.filter(
+            (nodeModularPipeline) =>
+              nodeModularPipeline.id === modularPipeline.id
+          )[0]
+        );
+      } else if (
+        modularPipelineParent.some(
+          (modularPipelineParentPipeline) =>
+            modularPipelineParentPipeline.id === modularPipeline.id
+        )
+      ) {
+        finalModularPipelines.push(
+          modularPipelineParent.filter(
+            (modularPipelineParentPipeline) =>
+              modularPipelineParentPipeline.id === modularPipeline.id
+          )[0]
+        );
+      }
+    });
+
+    return finalModularPipelines;
+  }
+);
+
+/**
+ * returns an array of the corresponding filtered nodes
+ * & unfiltered nodes for each filtered modular pipeline
+ */
+export const getFilteredModularPipelineNodes = createSelector(
+  [
+    getFilteredNodeItems,
+    getFilteredTreeItems,
+    (state) => state.modularPipelineIds,
+    (state) => state.nodeTypeIDs,
+  ],
+  (filteredNodeItems, filteredTreeItems, modularPipelineIDs, nodeTypeIDs) => {
+    const modularPipelineNodes = arrayToObject(modularPipelineIDs, () => []);
+
+    const nodeItems = cloneDeep(filteredNodeItems);
```

Comment:
Is there away to avoid `cloneDeep`? It's generally not a great idea for performance and can lead to data integrity bugs / confusion, so maybe there's another way?

Reply:
There is a need to use deepClone here actually to avoid bugs down the line - the item referenced here for `filteredNodeItems` will be used in other selectors later on and hence requires immutability. I had previously used shallow clone methods such as `Object.assign` and there proved to be problems down the line with the other selectors. 

So in short, deep clone is neccessary here. The approach and preference to use `cloneDeep` from lodash in our codebase was also advocated by Richard in his previous PRs, hence I am using it in this instance.

Reply:
Okay cool got it, thanks for clarifying.